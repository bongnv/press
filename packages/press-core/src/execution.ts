import path from "path";
import type { Hook } from "tapable";
import { AsyncParallelHook, AsyncSeriesHook } from "tapable";
import WebpackConfig from "webpack-chain";
import { Options } from "html-minifier";

export type PluginOptions = undefined | Record<string, any>;
export type Plugin = (execution: Execution, options: PluginOptions) => void;
type PluginNoOptions = string | Plugin;
export type Plugable =
  | [PluginNoOptions, PluginOptions]
  | [PluginNoOptions]
  | PluginNoOptions;

export interface Config {
  baseDir?: string;
  plugins?: Plugable[];
}

interface ExecutionHooks {
  prepare: Hook;
  configWebpack: Hook;
  bundle: Hook;
  generate: Hook;
}

interface ExecutionCommands {
  dev: Hook;
  build: Hook;
}

export class Execution {
  baseDir: string;
  outputDir: string;
  isProd: boolean;
  publicPath: string;
  temporaryPath: string;
  serverPath: string;
  htmlMinifierOptions?: Options;
  enhancers: string[];

  commands: ExecutionCommands;
  hooks: ExecutionHooks;

  clientWebpackConfig: WebpackConfig;
  serverWebpackConfig: WebpackConfig;

  urlsToRender: string[];

  constructor(config: Config) {
    this.baseDir = config.baseDir || process.cwd();
    this.outputDir = path.join(this.baseDir, "dist");
    this.isProd = process.env.NODE_ENv === "production";
    this.publicPath = "/";
    this.temporaryPath = path.join(this.baseDir, ".press");
    this.serverPath = path.join(this.temporaryPath, "server");

    const params = ["execution"];

    this.commands = {
      dev: new AsyncSeriesHook(params),
      build: new AsyncSeriesHook(params),
    };

    this.hooks = {
      prepare: new AsyncParallelHook(params),
      configWebpack: new AsyncSeriesHook(params),
      bundle: new AsyncParallelHook(params),
      generate: new AsyncParallelHook(params),
    };

    this.enhancers = [];
    this.clientWebpackConfig = new WebpackConfig();
    this.serverWebpackConfig = new WebpackConfig();
    this.urlsToRender = [];
  }
}
