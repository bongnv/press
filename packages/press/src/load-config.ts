import type { Config } from "@bongnv/press-core";
import vuePagesPlugin from "@bongnv/press-plugin-vue-pages";

export function loadConfig(argv: any): Config {
  return {
    baseDir: "/",
    plugins: [vuePagesPlugin],
  };
}
