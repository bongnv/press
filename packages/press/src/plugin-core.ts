import type WebpackConfig from "webpack-chain";
import type { Execution } from "./execution";
import type { Config } from "./load-config";

import path from "path";
import webpack from "webpack";
import MiniCssExtractPlugin from "mini-css-extract-plugin";
import { VueLoaderPlugin } from "vue-loader";
import { CleanWebpackPlugin } from "clean-webpack-plugin";
import VueSSRClientPlugin from "vue-server-renderer/client-plugin";
import VueServerBundlePlugin from "vue-server-renderer/server-plugin";
import nodeExternals from "webpack-node-externals";

const PLUGIN_NAME = "PressCore";

export class Core {
  vueAppDir: string;
  constructor() {
    this.vueAppDir = path.resolve(__dirname, "../vue-app");
  }

  apply({ commands, hooks }: Execution): void {
    hooks.configWebpack.tap(
      PLUGIN_NAME,
      ({ config, clientWebpackConfig, serverWebpackConfig }: Execution) => {
        this.applyClientConfig(config, clientWebpackConfig);
        this.applyServerConfig(config, serverWebpackConfig);
      },
    );

    commands.build.tap(PLUGIN_NAME, ({ hooks }: Execution) => {
      hooks.bundle.tapPromise(
        PLUGIN_NAME,
        ({ clientWebpackConfig, serverWebpackConfig }: Execution) =>
          Promise.all([
            this.runWebpack(clientWebpackConfig.toConfig()),
            this.runWebpack(serverWebpackConfig.toConfig()),
          ]),
      );
    });
  }

  private applyBaseConfig(
    isServer: boolean,
    isProd: boolean,
    webpackConfig: WebpackConfig,
  ) {
    webpackConfig.mode(isProd ? "production" : "development");

    webpackConfig.module
      .rule("compile-vue")
      .test(/\.vue$/)
      .use("vue-loader")
      .loader("vue-loader");

    webpackConfig.module
      .rule("compile-css")
      .test(/\.css$/)
      .use("mini-css-extract")
      .loader(MiniCssExtractPlugin.loader)
      .end()
      .use("css-loader")
      .loader("css-loader")
      .options({
        importLoaders: 1,
      })
      .end()
      .use("postcss-loader")
      .loader("postcss-loader")
      .end();

    webpackConfig.plugin("vue-loader").use(new VueLoaderPlugin());

    webpackConfig.plugin("mini-css-extract").use(
      new MiniCssExtractPlugin({
        filename: path.join(
          "_assets",
          isProd ? "[name].[contenthash].css" : "[name].css",
        ),
      }),
    );

    webpackConfig
      .plugin("clean")
      .use(new CleanWebpackPlugin({ cleanStaleWebpackAssets: false }));

    webpackConfig.node.set("setImmediate", false);

    webpackConfig.plugin("define").use(
      new webpack.DefinePlugin({
        "process.env.NODE_ENV": isProd ? '"production"' : '"development"',
        "process.env.VUE_ENV": isServer ? '"server"' : '"client"',
      }),
    );
  }

  private applyClientConfig(config: Config, webpackConfig: WebpackConfig) {
    const { isProd } = config;

    this.applyBaseConfig(false, isProd, webpackConfig);

    webpackConfig
      .entry("app")
      .add(path.join(this.vueAppDir, "entry-client.js"));

    webpackConfig.output
      .path(config.outputDir)
      .publicPath(config.publicPath)
      .filename(isProd ? "[name].[contenthash].js" : "[name].js")
      .chunkFilename(isProd ? "[name].[contenthash].js" : "[name].js");

    webpackConfig.plugin("vue-ssr-client").use(new VueSSRClientPlugin());
  }

  private applyServerConfig(config: Config, webpackConfig: WebpackConfig) {
    const { isProd, serverPath } = config;

    this.applyBaseConfig(true, isProd, webpackConfig);

    webpackConfig
      .target("node")
      .entry("app")
      .add(path.join(this.vueAppDir, "entry-server.js"))
      .end();

    webpackConfig.output.libraryTarget("commonjs2").path(serverPath);

    webpackConfig.plugin("vue-server-bundle").use(new VueServerBundlePlugin());
    webpackConfig.optimization.minimize(false);
    webpackConfig.externals(
      nodeExternals({
        // do not externalize CSS files in case we need to import it from a dep
        whitelist: /\.css$/,
      }),
    );
  }

  private runWebpack(config: webpack.Configuration): Promise<webpack.Stats> {
    return new Promise((resolve, reject) => {
      webpack(config, (err, stats) => {
        if (err) {
          console.error("error with webpack", err);
          return reject(err);
        }

        if (stats.hasErrors()) {
          console.error(stats.toString());
          return reject(new Error("Failed to compile with errors"));
        }

        console.log(stats.toString());
        resolve();
      });
    });
  }
}
