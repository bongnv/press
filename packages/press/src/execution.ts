import type { Hook } from "tapable";
import { AsyncParallelHook, AsyncSeriesHook } from "tapable";
import WebpackConfig from "webpack-chain";

import type { Config } from "./load-config";
import { Core } from "./plugin-core";

interface Plugin {
  // TODO: we may support async apply
  apply(execution: Execution): void;
}

interface ExecutionHooks {
  prepare: Hook;
  configWebpack: Hook;
  bundle: Hook;
  generate: Hook;
}

interface ExecutionCommands {
  dev: Hook;
  build: Hook;
}

export class Execution {
  config: Config;
  commands: ExecutionCommands;
  hooks: ExecutionHooks;

  clientWebpackConfig: WebpackConfig;
  serverWebpackConfig: WebpackConfig;

  constructor(config: Config) {
    this.config = config;

    // setup hooks
    const params = ["execution"];
    this.commands = {
      dev: new AsyncSeriesHook(params),
      build: new AsyncSeriesHook(params),
    };

    this.hooks = {
      prepare: new AsyncParallelHook(params),
      configWebpack: new AsyncSeriesHook(params),
      bundle: new AsyncParallelHook(params),
      generate: new AsyncParallelHook(params),
    };

    this.clientWebpackConfig = new WebpackConfig();
    this.serverWebpackConfig = new WebpackConfig();
  }

  async execBuild(): Promise<void> {
    this._applyPlugins();
    await this.commands.build.promise(this);
    await this.hooks.prepare.promise(this);
    await this.hooks.configWebpack.promise(this);
    await this.hooks.bundle.promise(this);
    await this.hooks.generate.promise(this);
  }

  private _applyPlugins(plugins: Plugin[] = []) {
    plugins.unshift(new Core());
    for (const plugin of plugins) {
      plugin.apply(this);
    }
  }
}
