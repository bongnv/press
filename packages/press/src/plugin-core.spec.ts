import MiniCssExtractPlugin from "mini-css-extract-plugin";

import type { Config } from "./load-config";
import { defaultConfig } from "./load-config";
import { Core } from "./plugin-core";
import { Execution } from "./execution";

jest.mock("mini-css-extract-plugin");

test("Core should apply configWebpack properly", async () => {
  MiniCssExtractPlugin.loader = "mini-css-extract-plugin-loader";

  const config: Config = defaultConfig("/");

  const corePlugin = new Core();
  corePlugin.vueAppDir = "/";
  const execution = new Execution(config);
  corePlugin.apply(execution);
  await execution.hooks.configWebpack.promise(execution);
  expect(execution.clientWebpackConfig.toConfig()).toMatchSnapshot();
  expect(execution.serverWebpackConfig.toConfig()).toMatchSnapshot();
});
