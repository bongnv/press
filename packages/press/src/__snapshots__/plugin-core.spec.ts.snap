// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core should apply configWebpack properly 1`] = `
Object {
  "entry": Object {
    "app": Array [
      "/entry-client.js",
    ],
  },
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\\\\\.vue\\$/,
        "use": Array [
          Object {
            "loader": "vue-loader",
          },
        ],
      },
      Object {
        "test": /\\\\\\.css\\$/,
        "use": Array [
          Object {
            "loader": "mini-css-extract-plugin-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
            },
          },
          Object {
            "loader": "postcss-loader",
          },
        ],
      },
    ],
  },
  "node": Object {
    "setImmediate": false,
  },
  "output": Object {
    "chunkFilename": "[name].js",
    "filename": "[name].js",
    "path": "/dist",
    "publicPath": "/",
  },
  "plugins": Array [
    VueLoaderPlugin {},
    MiniCssExtractPlugin {
      "apply": [MockFunction],
      "getCssChunkObject": [MockFunction],
      "renderContentAsset": [MockFunction],
    },
    CleanWebpackPlugin {
      "apply": [Function],
      "cleanAfterEveryBuildPatterns": Array [],
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
      ],
      "cleanStaleWebpackAssets": false,
      "currentAssets": Array [],
      "dangerouslyAllowCleanPatternsOutsideProject": false,
      "dry": false,
      "handleDone": [Function],
      "handleInitial": [Function],
      "initialClean": false,
      "outputPath": "",
      "protectWebpackAssets": true,
      "removeFiles": [Function],
      "verbose": false,
    },
    DefinePlugin {
      "definitions": Object {
        "process.env.NODE_ENV": "\\"development\\"",
        "process.env.VUE_ENV": "\\"client\\"",
      },
    },
    VueSSRClientPlugin {
      "options": Object {
        "filename": "vue-ssr-client-manifest.json",
      },
    },
  ],
}
`;

exports[`Core should apply configWebpack properly 2`] = `
Object {
  "entry": Object {
    "app": Array [
      "/entry-server.js",
    ],
  },
  "externals": [Function],
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\\\\\.vue\\$/,
        "use": Array [
          Object {
            "loader": "vue-loader",
          },
        ],
      },
      Object {
        "test": /\\\\\\.css\\$/,
        "use": Array [
          Object {
            "loader": "mini-css-extract-plugin-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
            },
          },
          Object {
            "loader": "postcss-loader",
          },
        ],
      },
    ],
  },
  "node": Object {
    "setImmediate": false,
  },
  "optimization": Object {
    "minimize": false,
  },
  "output": Object {
    "libraryTarget": "commonjs2",
    "path": "/.press/server",
  },
  "plugins": Array [
    VueLoaderPlugin {},
    MiniCssExtractPlugin {
      "apply": [MockFunction],
      "getCssChunkObject": [MockFunction],
      "renderContentAsset": [MockFunction],
    },
    CleanWebpackPlugin {
      "apply": [Function],
      "cleanAfterEveryBuildPatterns": Array [],
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
      ],
      "cleanStaleWebpackAssets": false,
      "currentAssets": Array [],
      "dangerouslyAllowCleanPatternsOutsideProject": false,
      "dry": false,
      "handleDone": [Function],
      "handleInitial": [Function],
      "initialClean": false,
      "outputPath": "",
      "protectWebpackAssets": true,
      "removeFiles": [Function],
      "verbose": false,
    },
    DefinePlugin {
      "definitions": Object {
        "process.env.NODE_ENV": "\\"development\\"",
        "process.env.VUE_ENV": "\\"server\\"",
      },
    },
    VueSSRServerPlugin {
      "options": Object {
        "filename": "vue-ssr-server-bundle.json",
      },
    },
  ],
  "target": "node",
}
`;
